File: js/conformance/runner.ts
import { readFile } from "node:fs/promises";
import type { BotFactory, BotUnderTest } from "./adapter.js";
import type { AdminFactory, AdminUnderTest } from "./admin-adapter.js";
import { FakeServer } from "./fake-server.js";
import { failure, matchHeaders, matchPartial, type Captures } from "./matchers.js";
import type { ConformanceCase, Failure, Json } from "./types.js";

/** Fixed test credential. Never a real token; cases refer to it as {token}. */
export const TEST_TOKEN = "bot:b1:" + "0".repeat(64);

const DEFAULT_TIMEOUT_MS = 5000;
/** After the exchanges run out, how long to watch for a request that should not come. */
const IDLE_GRACE_MS = 200;

export interface CaseResult {
  case: ConformanceCase;
  failures: Failure[];
}

export async function loadCase(path: string): Promise<ConformanceCase> {
  return JSON.parse(await readFile(path, "utf8")) as ConformanceCase;
}

export async function runCase(
  testCase: ConformanceCase,
  factory: BotFactory,
  managementFactory?: AdminFactory,
): Promise<CaseResult> {
  if (testCase.kind === "management") {
    if (!managementFactory) {
      return {
        case: testCase,
        failures: [failure("setup", "this case is a management case but no management factory was supplied")],
      };
    }
    return runManagementCase(testCase, managementFactory);
  }

  const server = new FakeServer(testCase.exchanges);
  await server.start();

  let bot: BotUnderTest | null = null;
  const failures: Failure[] = [];

  // A conformance runner executes untrusted code by definition: an SDK with a
  // bug is the whole point. An unhandled rejection from the bot under test must
  // fail THIS case and let the suite continue — otherwise one broken SDK hides
  // every case after it, which is exactly when the report matters most.
  let secondStartRejected = false;
  let stopDurationMs: number | undefined;
  let stopped = false;
  const crashes: string[] = [];
  const onCrash = (reason: unknown) => crashes.push(describe(reason));
  process.on("unhandledRejection", onCrash);
  process.on("uncaughtException", onCrash);

  try {
    bot = factory.create({
      baseUrl: server.baseUrl,
      token: TEST_TOKEN,
      limit: testCase.bot?.options?.limit,
      timeoutSeconds: testCase.bot?.options?.timeoutSeconds,
      handler: testCase.bot?.handler ?? { kind: "noop" },
    });
    await bot.start();

    if (testCase.bot?.startTwice) {
      // G1: the second start must fail locally. If it silently succeeds, the
      // SDK has two loops on one bot and is causing the very 409 it should be
      // reporting.
      try {
        await bot.start();
      } catch {
        secondStartRejected = true;
      }
    }

    const stopAfterMs = testCase.run?.stopAfterMs;
    if (stopAfterMs !== undefined) {
      await sleep(stopAfterMs);
      const startedStopping = Date.now();
      await bot.stop();
      stopDurationMs = Date.now() - startedStopping;
      stopped = true;
    } else {
      await waitForCompletion(bot, server, testCase.run?.timeoutMs ?? DEFAULT_TIMEOUT_MS);
    }
  } catch (error) {
    failures.push(failure("run", `the bot threw out of start(): ${describe(error)}`));
  } finally {
    try {
      if (!stopped) await bot?.stop();
    } catch {
      /* stopping must never mask a real failure */
    }
    await server.stop();
    process.off("unhandledRejection", onCrash);
    process.off("uncaughtException", onCrash);
  }

  for (const crash of crashes) {
    failures.push(failure("crash", `the bot threw and did not handle it: ${crash}`));
  }
  failures.push(...checkRequests(testCase, server));
  if (bot) failures.push(...checkAssertions(testCase, bot, { secondStartRejected, stopDurationMs }));
  return { case: testCase, failures };
}

/**
 * A management case invokes methods in order; nothing polls, so there is no
 * loop to wait on and no steady state. The exchange list and every matcher work
 * exactly as they do for the runtime — that reuse is why the two kinds share a
 * format at all.
 */
async function runManagementCase(
  testCase: ConformanceCase,
  factory: AdminFactory,
): Promise<CaseResult> {
  const server = new FakeServer(testCase.exchanges);
  await server.start();
  const failures: Failure[] = [];
  let client: AdminUnderTest | null = null;
  try {
    client = factory.create({ baseUrl: server.baseUrl });
    for (const call of testCase.calls ?? []) {
      await client.invoke(call.method, (call.args ?? {}) as Record<string, unknown>);
    }
  } catch (error) {
    failures.push(failure("run", `the client threw out of a call: ${describe(error)}`));
  } finally {
    await server.stop();
  }

  failures.push(...checkRequests(testCase, server));
  if (client) failures.push(...checkManagementAssertions(testCase, client));
  return { case: testCase, failures };
}

function checkManagementAssertions(testCase: ConformanceCase, client: AdminUnderTest): Failure[] {
  const failures: Failure[] = [];
  const want = testCase.assert;
  if (!want) return failures;

  for (const [i, expected] of (want.errorsReported ?? []).entries()) {
    const actual = client.errors[i];
    if (!actual) {
      failures.push(failure("assert.errorsReported", `expected an error at position ${i}, none was reported`));
      continue;
    }
    if (expected.managementCode !== undefined && actual.managementCode !== expected.managementCode) {
      failures.push(
        failure(
          "assert.errorsReported",
          `error #${i + 1}: expected code ${expected.managementCode}, got ${actual.managementCode}`,
        ),
      );
    }
    for (const field of ["retryable", "accessLost"] as const) {
      if (expected[field] !== undefined && actual[field] !== expected[field]) {
        failures.push(
          failure(
            "assert.errorsReported",
            `error #${i + 1}: expected ${field}=${expected[field]}, got ${actual[field]}`,
          ),
        );
      }
    }
  }
  if ((want.errorsReported?.length ?? 0) === 0 && client.errors.length > 0) {
    failures.push(
      failure("assert.errorsReported", `expected no errors, got ${JSON.stringify(client.errors.map((e) => e.managementCode ?? e.message))}`),
    );
  }

  if (want.createdBotId !== undefined) {
    const reported = client.results.some(
      (result) => (result as { botId?: string } | null)?.botId === want.createdBotId,
    );
    if (!reported) {
      failures.push(
        failure("assert.createdBotId", `expected the facade to report creating ${want.createdBotId}`),
      );
    }
  }

  if (want.resultContains !== undefined) {
    // Values, never field names: names are idiomatic per language.
    const encoded = JSON.stringify(client.results);
    for (const value of want.resultContains) {
      if (!encoded.includes(value)) {
        failures.push(failure("assert.resultContains", `the value ${JSON.stringify(value)} never reached the caller`));
      }
    }
  }

  if (want.secretReturnedOnce !== undefined) {
    const returned = client.results.some((result) => JSON.stringify(result ?? null).includes(want.secretReturnedOnce!));
    if (!returned) {
      failures.push(failure("assert.secretReturnedOnce", "the revealed token never reached the caller"));
    }
    // The one response carrying a live token is response-only by the server's
    // own design. If the SDK lets it reach a log line, an error, or its own
    // representation, it undoes the only protection built for it.
    const places: Record<string, string> = {
      errors: JSON.stringify(client.errors),
      warnings: JSON.stringify(client.warnings),
      repr: client.describe(),
    };
    for (const place of want.secretNotIn ?? []) {
      if (places[place]?.includes(want.secretReturnedOnce)) {
        failures.push(
          failure("assert.secretNotIn", `the revealed token LEAKED into ${place}; it must reach the return value only`),
        );
      }
    }
  }
  return failures;
}

async function waitForCompletion(bot: BotUnderTest, server: FakeServer, timeoutMs: number): Promise<void> {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    if (bot.stoppedItself) {
      // Even a stopped bot gets watched briefly: a bot that reports the error
      // and keeps polling is exactly the g7-409 failure this suite exists for.
      await sleep(IDLE_GRACE_MS);
      return;
    }
    if (server.exhausted) {
      await sleep(IDLE_GRACE_MS);
      return;
    }
    await sleep(10);
  }
}

function checkRequests(testCase: ConformanceCase, server: FakeServer): Failure[] {
  const failures: Failure[] = [];
  const captures: Captures = new Map();
  /** Per-request problems, collapsed at the end so one defect reads as one line. */
  const perRequest: { index: number; detail: string }[] = [];
  const note = (index: number, detail: string) => perRequest.push({ index, detail });

  const last = testCase.exchanges[testCase.exchanges.length - 1];
  const repeats = last?.repeat === true;
  // With a repeating tail, every poll past the declared list must still match
  // that last expectation — the bot may idle, but it may not idle differently.
  const total = repeats ? Math.max(testCase.exchanges.length, server.observed.length) : testCase.exchanges.length;

  for (let i = 0; i < total; i++) {
    const expected = (testCase.exchanges[i] ?? last)!.expect;
    const actual = server.observed[i];
    if (!actual) {
      note(i + 1, `expected ${describeExpected(expected)}, but the SDK never sent it`);
      continue;
    }
    if (expected.method && expected.method !== actual.method) {
      note(i + 1, `expected method ${expected.method}, got ${actual.method}`);
    }
    // The observed path carries the query string; the case declares them
    // separately so a query can be partially matched like a body.
    const [actualPath, actualQuery = ""] = actual.path.split("?");
    if (expected.path) {
      const want = expected.path.replaceAll("{token}", TEST_TOKEN);
      if (want !== actualPath) {
        note(i + 1, `expected path ${redact(want)}, got ${redact(actualPath!)}`);
      }
    }
    if (expected.query) {
      const observed: Record<string, Json> = {};
      for (const [key, value] of new URLSearchParams(actualQuery)) observed[key] = value;
      const reason = matchPartial(expected.query as Record<string, Json>, observed, captures, "query");
      if (reason) note(i + 1, reason);
    }
    if (expected.headers) {
      const reason = matchHeaders(expected.headers, actual.headers, captures);
      if (reason) note(i + 1, reason);
    }
    if (expected.body) {
      const reason = matchPartial(expected.body as Record<string, Json>, actual.body, captures, "body");
      if (reason) note(i + 1, reason);
    }
  }

  const byDetail = new Map<string, number[]>();
  for (const { index, detail } of perRequest) {
    (byDetail.get(detail) ?? byDetail.set(detail, []).get(detail)!).push(index);
  }
  for (const [detail, indexes] of byDetail) {
    const where =
      indexes.length === 1
        ? `request #${indexes[0]}`
        : `requests #${indexes[0]}-#${indexes[indexes.length - 1]} (${indexes.length} times)`;
    failures.push(failure(where, detail));
  }

  if (server.extras.length > 0) {
    const shown = server.extras.slice(0, 3).map((e) => `${e.method} ${redact(e.path)}`);
    const rest = server.extras.length - shown.length;
    failures.push(
      failure(
        "unexpected requests",
        `the SDK sent ${server.extras.length} request(s) after the last declared exchange ` +
          `(${shown.join(", ")}${rest > 0 ? `, +${rest} more` : ""}). ` +
          `The exchange list is exhaustive: nothing may follow it.`,
      ),
    );
  }
  return failures;
}

interface Lifecycle {
  secondStartRejected: boolean;
  stopDurationMs: number | undefined;
}

function checkAssertions(testCase: ConformanceCase, bot: BotUnderTest, lifecycle: Lifecycle): Failure[] {
  const failures: Failure[] = [];
  const want = testCase.assert;
  if (!want) return failures;

  if (want.secondStartRejected !== undefined && want.secondStartRejected !== lifecycle.secondStartRejected) {
    failures.push(
      failure(
        "assert.secondStartRejected",
        want.secondStartRejected
          ? "the second start() succeeded; the SDK now has two polls on one bot"
          : "the second start() was rejected, but this case expected it to be allowed",
      ),
    );
  }

  if (want.stoppedWithinMs !== undefined) {
    if (lifecycle.stopDurationMs === undefined) {
      failures.push(
        failure("assert.stoppedWithinMs", "this case asserts on stop() but never called it; add run.stopAfterMs"),
      );
    } else if (lifecycle.stopDurationMs > want.stoppedWithinMs) {
      failures.push(
        failure(
          "assert.stoppedWithinMs",
          `stop() took ${lifecycle.stopDurationMs}ms, over the ${want.stoppedWithinMs}ms budget — ` +
            `it is waiting out the server's response instead of aborting the request`,
        ),
      );
    }
  }

  if (want.botStopped !== undefined && want.botStopped !== bot.stoppedItself) {
    failures.push(
      failure(
        "assert.botStopped",
        want.botStopped
          ? "expected the bot to stop on its own, but it was still running"
          : "expected the bot to keep running, but it stopped on its own",
      ),
    );
  }

  for (const run of want.handlerRuns ?? []) {
    const actual = bot.handlerRuns.get(run.updateId) ?? 0;
    if (actual !== run.times) {
      failures.push(
        failure("assert.handlerRuns", `update ${run.updateId}: expected ${run.times} run(s), got ${actual}`),
      );
    }
  }

  for (const [i, expected] of (want.errorsReported ?? []).entries()) {
    const actual = bot.errors[i];
    if (!actual) {
      failures.push(failure("assert.errorsReported", `expected an error at position ${i}, none was reported`));
      continue;
    }
    if (expected.code !== undefined && actual.code !== expected.code) {
      failures.push(
        failure("assert.errorsReported", `error #${i + 1}: expected code ${expected.code}, got ${actual.code}`),
      );
    }
    if (expected.notContains) {
      const needle = expected.notContains.replaceAll("{token}", TEST_TOKEN);
      if (actual.message.includes(needle)) {
        failures.push(
          failure(
            "assert.errorsReported",
            `error #${i + 1} LEAKS the token. An error crossing the SDK boundary must be redacted first.`,
          ),
        );
      }
    }
  }

  for (const warning of want.warnings ?? []) {
    if (!bot.warnings.some((w) => w.toLowerCase().includes(warning.contains.toLowerCase()))) {
      failures.push(
        failure("assert.warnings", `expected a warning containing ${JSON.stringify(warning.contains)}, got ${
          bot.warnings.length ? JSON.stringify(bot.warnings) : "none"
        }`),
      );
    }
  }
  return failures;
}

function describeExpected(expected: { method?: string; path?: string }): string {
  return `${expected.method ?? "POST"} ${redact(expected.path ?? "?")}`;
}

/** The runner prints paths, and paths carry the token. Redact before printing. */
function redact(text: string): string {
  return text.replaceAll(TEST_TOKEN, "<TOKEN>").replace(/\/bot[^/]+\//, "/bot<TOKEN>/");
}

function describe(error: unknown): string {
  return redact(error instanceof Error ? error.message : String(error));
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: go/conformance/runner.go
package conformance

import (
	"context"
	"fmt"
	"net/url"
	"regexp"
	"strings"
	"time"
)

// TestToken is a fixed credential. Never a real token; cases refer to it as
// {token}.
const TestToken = "bot:b1:" + "0000000000000000000000000000000000000000000000000000000000000000"

const defaultTimeout = 5 * time.Second

// After the exchanges run out, how long to watch for a request that should not
// come.
const idleGrace = 200 * time.Millisecond

type Failure struct {
	Where  string
	Detail string
}

func (f Failure) String() string { return f.Where + ": " + f.Detail }

// RunCase executes one runtime case and returns everything that went wrong.
// Management cases go through RunManagementCase instead.
func RunCase(c Case, factory Factory) []Failure {
	if c.Kind == "management" {
		return runManagementCase(c, NewSDKManagement)
	}
	fake := NewFakeServer(c.Exchanges)
	defer fake.Close()

	bot := factory(BotOptions{
		BaseURL:        fake.BaseURL(),
		Token:          TestToken,
		Limit:          c.Bot.Options.Limit,
		TimeoutSeconds: c.Bot.Options.TimeoutSeconds,
		Handler:        c.Bot.Handler,
	})

	timeout := defaultTimeout
	if c.Run.TimeoutMs > 0 {
		timeout = time.Duration(c.Run.TimeoutMs) * time.Millisecond
	}
	ctx, cancel := context.WithCancel(context.Background())
	bot.Start(ctx)

	secondStartRejected := false
	if c.Bot.StartTwice {
		// Give the first Run a moment to take the loop, so the second call is
		// answering "already running" rather than winning a race.
		time.Sleep(50 * time.Millisecond)
		secondStartRejected = bot.StartAgain() != nil
	}

	stopDuration := time.Duration(-1)
	if c.Run.StopAfterMs > 0 {
		time.Sleep(time.Duration(c.Run.StopAfterMs) * time.Millisecond)
		startedStopping := time.Now()
		cancel()
		bot.Stop()
		stopDuration = time.Since(startedStopping)
	} else {
		waitForCompletion(bot, fake, timeout)
		cancel()
		bot.Stop()
	}

	failures := checkRequests(c, fake)
	failures = append(failures, checkAssertions(c, bot, secondStartRejected, stopDuration)...)
	return failures
}

// RunManagementCase drives a case against a specific management client, so a
// test can point it at a deliberately broken one and prove this runner checks.
func RunManagementCase(c Case, factory ManagementFactory) []Failure {
	return runManagementCase(c, factory)
}

func waitForCompletion(bot BotUnderTest, fake *FakeServer, timeout time.Duration) {
	deadline := time.Now().Add(timeout)
	for time.Now().Before(deadline) {
		if bot.StoppedItself() {
			// Even a stopped bot is watched briefly: one that reports the error
			// and keeps polling is exactly the g7-409 failure this suite exists
			// to catch.
			time.Sleep(idleGrace)
			return
		}
		if fake.Exhausted() {
			time.Sleep(idleGrace)
			return
		}
		time.Sleep(10 * time.Millisecond)
	}
}

func checkRequests(c Case, fake *FakeServer) []Failure {
	var failures []Failure
	captures := Captures{}
	observed := fake.Observed()

	// Per-request problems, collapsed at the end so one defect reads as one
	// line: the idle steady state repeats the same poll many times.
	type problem struct {
		index  int
		detail string
	}
	var problems []problem
	note := func(index int, detail string) { problems = append(problems, problem{index, detail}) }

	repeats := len(c.Exchanges) > 0 && c.Exchanges[len(c.Exchanges)-1].Repeat
	total := len(c.Exchanges)
	if repeats && len(observed) > total {
		total = len(observed)
	}

	for i := 0; i < total; i++ {
		expected := c.Exchanges[min(i, len(c.Exchanges)-1)].Expect
		if i >= len(observed) {
			note(i+1, fmt.Sprintf("expected %s %s, but the SDK never sent it",
				orDefault(expected.Method, "POST"), redact(expected.Path)))
			continue
		}
		actual := observed[i]
		if expected.Method != "" && expected.Method != actual.Method {
			note(i+1, fmt.Sprintf("expected method %s, got %s", expected.Method, actual.Method))
		}
		// The observed path carries the query string; the case declares them
		// separately so a query can be partially matched like a body.
		actualPath, actualQuery, _ := strings.Cut(actual.Path, "?")
		if expected.Path != "" {
			want := strings.ReplaceAll(expected.Path, "{token}", TestToken)
			if want != actualPath {
				note(i+1, fmt.Sprintf("expected path %s, got %s", redact(want), redact(actualPath)))
			}
		}
		if len(expected.Query) > 0 {
			observed := map[string]any{}
			if values, err := url.ParseQuery(actualQuery); err == nil {
				for key := range values {
					observed[key] = values.Get(key)
				}
			}
			if reason := matchPartial(expected.Query, observed, captures, "query"); reason != "" {
				note(i+1, reason)
			}
		}
		if len(expected.Headers) > 0 {
			if reason := matchHeaders(expected.Headers, actual.Headers, captures); reason != "" {
				note(i+1, reason)
			}
		}
		if len(expected.Body) > 0 {
			if reason := matchPartial(expected.Body, actual.Body, captures, "body"); reason != "" {
				note(i+1, reason)
			}
		}
	}

	seen := map[string][]int{}
	var order []string
	for _, p := range problems {
		if _, ok := seen[p.detail]; !ok {
			order = append(order, p.detail)
		}
		seen[p.detail] = append(seen[p.detail], p.index)
	}
	for _, detail := range order {
		indexes := seen[detail]
		where := fmt.Sprintf("request #%d", indexes[0])
		if len(indexes) > 1 {
			where = fmt.Sprintf("requests #%d-#%d (%d times)", indexes[0], indexes[len(indexes)-1], len(indexes))
		}
		failures = append(failures, Failure{Where: where, Detail: detail})
	}

	if extras := fake.Extras(); len(extras) > 0 {
		shown := make([]string, 0, 3)
		for _, e := range extras {
			if len(shown) == 3 {
				break
			}
			shown = append(shown, e.Method+" "+redact(e.Path))
		}
		more := ""
		if rest := len(extras) - len(shown); rest > 0 {
			more = fmt.Sprintf(", +%d more", rest)
		}
		failures = append(failures, Failure{
			Where: "unexpected requests",
			Detail: fmt.Sprintf("the SDK sent %d request(s) after the last declared exchange (%s%s). "+
				"The exchange list is exhaustive: nothing may follow it.", len(extras), strings.Join(shown, ", "), more),
		})
	}
	return failures
}

func checkAssertions(c Case, bot BotUnderTest, secondStartRejected bool, stopDuration time.Duration) []Failure {
	var failures []Failure

	if want := c.Assert.SecondStartRejected; want != nil && *want != secondStartRejected {
		detail := "the second Start was rejected, but this case expected it to be allowed"
		if *want {
			detail = "the second Start succeeded; the SDK now has two polls on one bot"
		}
		failures = append(failures, Failure{Where: "assert.secondStartRejected", Detail: detail})
	}

	if want := c.Assert.StoppedWithinMs; want != nil {
		switch {
		case stopDuration < 0:
			failures = append(failures, Failure{
				Where:  "assert.stoppedWithinMs",
				Detail: "this case asserts on Stop but never called it; add run.stopAfterMs",
			})
		case stopDuration > time.Duration(*want)*time.Millisecond:
			failures = append(failures, Failure{
				Where: "assert.stoppedWithinMs",
				Detail: fmt.Sprintf("Stop took %dms, over the %dms budget — it is waiting out the "+
					"server's response instead of cancelling the request", stopDuration.Milliseconds(), *want),
			})
		}
	}

	if want := c.Assert.BotStopped; want != nil && *want != bot.StoppedItself() {
		detail := "expected the bot to keep running, but it stopped on its own"
		if *want {
			detail = "expected the bot to stop on its own, but it was still running"
		}
		failures = append(failures, Failure{Where: "assert.botStopped", Detail: detail})
	}

	runs := bot.HandlerRuns()
	for _, want := range c.Assert.HandlerRuns {
		if got := runs[want.UpdateID]; got != want.Times {
			failures = append(failures, Failure{
				Where:  "assert.handlerRuns",
				Detail: fmt.Sprintf("update %d: expected %d run(s), got %d", want.UpdateID, want.Times, got),
			})
		}
	}

	errs := bot.Errors()
	for i, want := range c.Assert.ErrorsReported {
		if i >= len(errs) {
			failures = append(failures, Failure{
				Where:  "assert.errorsReported",
				Detail: fmt.Sprintf("expected an error at position %d, none was reported", i),
			})
			continue
		}
		got := errs[i]
		if want.Code != nil && (!got.HasCode || got.Code != *want.Code) {
			failures = append(failures, Failure{
				Where:  "assert.errorsReported",
				Detail: fmt.Sprintf("error #%d: expected code %d, got %d", i+1, *want.Code, got.Code),
			})
		}
		if want.NotContains != "" {
			needle := strings.ReplaceAll(want.NotContains, "{token}", TestToken)
			if strings.Contains(got.Message, needle) {
				failures = append(failures, Failure{
					Where: "assert.errorsReported",
					Detail: fmt.Sprintf("error #%d LEAKS the token. An error crossing the SDK boundary "+
						"must be redacted first.", i+1),
				})
			}
		}
	}

	warnings := bot.Warnings()
	for _, want := range c.Assert.Warnings {
		found := false
		for _, w := range warnings {
			if strings.Contains(strings.ToLower(w), strings.ToLower(want.Contains)) {
				found = true
				break
			}
		}
		if !found {
			failures = append(failures, Failure{
				Where:  "assert.warnings",
				Detail: fmt.Sprintf("expected a warning containing %q, got %v", want.Contains, warnings),
			})
		}
	}
	return failures
}

var botPathPattern = regexp.MustCompile(`/bot[^/\s]+`)

// The runner prints paths, and paths carry the token.
func redact(text string) string {
	return botPathPattern.ReplaceAllString(strings.ReplaceAll(text, TestToken, "<TOKEN>"), "/bot<TOKEN>")
}

func orDefault(value, fallback string) string {
	if value == "" {
		return fallback
	}
	return value
}
